### PR TITLE
Migrated selectors to return elementList instead of elementWrapper

### PR DESCRIPTION
**Pull request by luciferankon**

### CODE REVIEW

**Comment by nehashri** [comment]
@luciferankon Please resolve the conflicts on this PR.

**Comment by luciferankon** [comment]
> @luciferankon Please resolve the conflicts on this PR.

@nehashri done

File: lib/taiko.js [hunk 1]
```diff
@@ -2045,11 +2072,35 @@ module.exports.image = (attrValuePairs, ...args) => {
...
     '//img|//*[contains(@style, "background-image")]',
   );
+  const description = desc(selector, 'alt', 'image');
+
   return {
-    get: getIfExists(get),
-    exists: exists(getIfExists(get), descEvent),
-    description: desc(selector, 'alt', 'Image'),
-    text: selectorText(get),
+    get: async (tag, retryInterval, retryTimeout) => {
```
Comment (nehashri) [inline lib/taiko.js]: Can we extract the common method into another method?

File: lib/helper.js [hunk 2]
```diff
@@ -146,9 +148,43 @@ const getIfExists = get => {
...
+            );
+            return result.result.value;
+          },
+          description,
+        };
+      });
+      return elementWrappers;
     } catch (e) {
-      return [];
+      return [
```
Comment (nehashri) [inline lib/helper.js]: We cannot return an array with a dummy object here. The `.elements` api will not work properly if no element is found.
Eg. if element does not exist, `.elements.length` is not equal to 0.

File: lib/helper.js [hunk 3]
```diff
@@ -146,9 +148,43 @@ const getIfExists = get => {
...
+              null,
+              { nodeId },
+            );
+            return result.result.value;
+          },
+          description,
+        };
+      });
+      return elementWrappers;
     } catch (e) {
```
Comment (nehashri) [inline lib/helper.js]: Please log the exception.

File: lib/helper.js [hunk 4]
```diff
@@ -198,6 +229,41 @@ const handleUrlRedirection = url => {
...
+      return elems[0].exists();
+    },
+    description,
+    text: async (tag, retryInterval, retryTimeout) => {
+      const elems = await getIfExists(get, description, descEvent)(
+        tag,
+        retryInterval,
+        retryTimeout,
+      );
+      return await elems[0].text();
```
Comment (nehashri) [inline lib/helper.js]: What will `.text()` return if no element exists?

File: test/unit-tests/$.test.js [hunk 5]
```diff
@@ -40,14 +40,72 @@ describe(test_name, () => {
...
+    });
+
+    it('test description', async () => {
+      const elems = await $('#foo').elements();
+      expect(elems[0].description).to.be.eql(
+        'Custom selector $(#foo)',
+      );
+    });
+
+    it('test text()', async () => {
```
Comment (nehashri) [inline test/unit-tests/$.test.js]: write test for failing case

File: lib/taiko.js [hunk 6]
```diff
@@ -3566,27 +3673,15 @@ const validate = () => {
...
-    const r = await domHandler.getBoundingClientRect(e);
-    results.push({ elem: e, result: callback(r) });
+    if (e.get) {
+      const nodeId = await e.get();
+      const r = await domHandler.getBoundingClientRect(nodeId);
+      results.push({ elem: nodeId, result: callback(r) });
+    }
   }
   return results;
 };
```
Comment (NivedhaSenthil) [inline lib/taiko.js]: update docs for element wrapper

File: lib/helper.js [hunk 7]
```diff
@@ -146,7 +151,29 @@ const getIfExists = get => {
...
+      const nodeIds = await get(tag);
+      function getText() {
+        return this.innerText;
+      }
+      let elementWrappers = nodeIds.map(nodeId => {
+        if (nodeId instanceof Object) {
+          return Object.assign(nodeId, { description: description });
+        }
+        let basicElementWrapper = {
+          get: async () => nodeId,
```
Comment (NivedhaSenthil) [inline lib/helper.js]: this can be a simple value instead of a func as it returns just the nodeId

Comment (nehashri) [reply lib/helper.js]: done

File: lib/taiko.js [hunk 8]
```diff
@@ -2271,9 +2264,11 @@ module.exports.inputField = (attrValuePairs, ...args) => {
       const nodeId = (await getIfExists(get)())[0];
```
[the remark this answers is not in the document]
Comment (nehashri) [reply lib/taiko.js]: done.

File: lib/taiko.js (hunk 6 above)
Comment (nehashri) [reply lib/taiko.js]: done

**Comment by nehashri** [comment]
@NivedhaSenthil Please recheck this PR. I have made the required changes. 

PS: Please `squash` this PR

**Pull request merged**